torch/csrc/jit/passes/canonicalize.cpp

```diff
@@ -47,5 +48,220 @@ std::shared_ptr<Graph> Canonicalize(
+  // Invalid to call with the same node as both args
+  AT_ASSERT(n1 != n2);
+
+  // Set n1 and n2 to be the number of blocks from the Graph block
+  size_t d_1 = blocksFromGraphBlock(n1);
+  size_t d_2 = blocksFromGraphBlock(n2);
+
+  for (; d_1 > d_2; --d_1) {
+    n1 = n1->owningBlock()->owningNode();
+    // n2 contains n1
+    if (n1 == n2) {
+      return false;
+    }
+  }
+
+  for (; d_2 > d_1; --d_2) {
+    n2 = n2->owningBlock()->owningNode();
+    // n1 contains n2
+    if (n2 == n1) {
+      return true;
+    }
+  }
+
+  // Now they are the same numer of blocks from the graph block,
+  // recurse upwards, checking if they are on the same block
+  while (true) {
+    if (n1->owningBlock() == n2->owningBlock()) {
+      return n1->isBefore(n2);
+    }
+
+    auto new_n1 = n1->owningBlock()->owningNode();
+    auto new_n2 = n2->owningBlock()->owningNode();
+
+    AT_ASSERT(new_n1 != nullptr);
+    AT_ASSERT(new_n2 != nullptr);
+
+    if (new_n1 == new_n2) {
+      // take whichever node is in the earlier block
+      auto index_1 = blockIndex(n1->owningBlock());
+      auto index_2 = blockIndex(n2->owningBlock());
+      return index_1 < index_2;
+    }
+
+    n1 = new_n1;
+    n2 = new_n2;
+  }
+}
+
+std::vector<size_t> sort_indexes(at::ArrayRef<Value*> values) {
+  // initialize original index locations
+  std::vector<size_t> idx(values.size());
+  std::iota(idx.begin(), idx.end(), 0);
+
+  // Sort values based on canonical ordering of their first usage
+  std::sort(idx.begin(), idx.end(), [&values](size_t i1, size_t i2) {
+    if (values[i1]->uses().size() == 0 && values[i2]->uses().size() == 0) {
+      return i1 < i2;
+    }
+    if (values[i1]->uses().size() == 0) {
+      return false;
+    } else if (values[i2]->uses().size() == 0) {
+      return true;
+    }
+
+    auto fst_v1 = values[i1]->uses()[0];
+    auto fst_v2 = values[i2]->uses()[0];
+
+    // If two values first usage is the same node, we order on offset
+    if (fst_v1.user == fst_v2.user) {
+      return fst_v1.offset < fst_v2.offset;
+    }
+
+    return isBefore(fst_v1.user, fst_v2.user);
+  });
+
+  return idx;
+}
+
+std::string uniqueName(Value* v) {
+  return v->hasUniqueName() ? v->uniqueName() : "";
+}
+
+void CanonicalizeOutputs(Block* block);
+
+void swapIfNodeOutputs(Node* n, const std::vector<size_t>& new_indices) {
+  for (size_t index : new_indices) {
+    auto orig = n->outputs().at(index);
+    auto new_out =
+        n->addOutput()->setUniqueName(uniqueName(orig))->setType(orig->type());
+    orig->replaceAllUsesWith(new_out);
+  }
+  while (n->outputs().size() > new_indices.size()) {
+    n->eraseOutput(0);
+  }
+}
+
+void swapIfBlockOutputs(Block* b, const std::vector<size_t>& new_indices) {
+  for (size_t index : new_indices) {
+    b->registerOutput(b->outputs().at(index));
+  }
+  for (size_t i = 0; i < new_indices.size(); ++i) {
+    b->eraseOutput(0);
+  }
+}
+
+void swapLoopBlockOutputs(Node* n, const std::vector<size_t>& new_indices) {
```

Comment:
This code is hard to follow because it is combining a couple things into one:

1. Permuting inputs or outputs to a function.
2. Keeping track of the extra non-permuted inputs

Instead can we:

1. add permuteInputs/permuteOutputs directly to ir.h (If you look at what removeInput does, it is not actually that hard, it just requires adjusting the Use objects offsets).  This avoids having all the logic to copy meta-data because we are just swapping the nodes
2. Add permuteLoopCarried and permuteOutputs to LoopView, IfView, because they are the only classes that should know about carried_inputs_offsets, etc.
3. Redo this code to just call the  view versions.